6_visualize/src/prep_spark_line_fun.R

```diff
@@ -0,0 +1,71 @@
+    y_space <- viz_config$sparks$ytop-y_start
+    vertical_spacing <- y_space/length(sites)
+
+    y_pos <- y_start # initialize position of first spark
+    for(i in sites) {
+
+      # Filter data to just one site & create polygon out of it
+      storm_data_i <- filter(this_spark, site_no == i) %>% arrange(dateTime)
+      storm_data_site_all <- filter(storm_data, site_no == i)
+
+      if(nrow(storm_data_i) == 0) { next } # skip if there's no data at or before this timestep
+
+      flood_stage_va <- as.numeric(unique(storm_data_i$flood_stage_normalized))
+      lowest_stage <- min(storm_data_site_all$stage_normalized, na.rm = TRUE)
+      highest_stage <- max(storm_data_site_all$stage_normalized, na.rm = TRUE)
+
+      full_polygon <- data.frame(dateTime = head(storm_data_i$dateTime, 1),
+                                 stage_normalized = min(storm_data_i$stage_normalized,1)) %>%
+        bind_rows(select(storm_data_i, dateTime, stage_normalized)) %>%
+        bind_rows(data.frame(dateTime = tail(storm_data_i$dateTime, 1),
+                             stage_normalized = min(storm_data_i$stage_normalized)))
+
+      # Fill values lower than flood stage with the stage & then create a polygon out of it
+      above_flood_data <- mutate(storm_data_i, stage_normalized = ifelse(stage_normalized < flood_stage_va,
+                                                                         flood_stage_va,
+                                                                         stage_normalized))
+      above_flood_polygon <- data.frame(dateTime = head(above_flood_data$dateTime, 1),
+                                        stage_normalized = head(above_flood_data$stage_normalized, 1)) %>%
+        bind_rows(select(above_flood_data, dateTime, stage_normalized)) %>%
+        bind_rows(data.frame(dateTime = tail(above_flood_data$dateTime, 1),
+                             stage_normalized = min(above_flood_data$stage_normalized)))
+
+      y_coords <- c(y_pos, y_pos + vertical_spacing) # y coords relative to plot box
+      fig.new <- c(grconvertX(x_coords, from="npc", to="ndc"), grconvertY(y_coords, from="npc", to="ndc"))
+
+      op <- par(fig=fig.new, cex=2, new=TRUE, mar=rep(0, 4)) # setup parameters
+      # setup a plotting device
+      plot(x=NA, y=NA, type='n', axes=FALSE, xlab="", ylab="",
+           xlim = c(as.POSIXct(viz_config$dates$start, tz = "UTC"), as.POSIXct(viz_config$dates$end, tz = "UTC")),
+           ylim = c(lowest_stage, highest_stage))
+      # put stage polygons on plot
+      polygon(full_polygon$dateTime, full_polygon$stage_normalized, col = viz_config$gage_norm_col, border=NA)
+      polygon(above_flood_polygon$dateTime, above_flood_polygon$stage_normalized, col = viz_config$gage_flood_col, border=NA)
+      points(full_polygon$dateTime[-nrow(full_polygon)], full_polygon$stage_normalized[-nrow(full_polygon)],
+             col = viz_config$gage_flood_col, type='l', lwd=2.5)
+      par(op) # reset plot parameters
```

Comment:
i see there are two layers of par() setting and resetting here, one within the loop and one outside it. why is this particular call needed - which plot parameters get reset here?

Reply:
Hmm I guess I only need to reset once within the loop.

```
plot(NA, NA, xlim=c(0,1), ylim=c(0,1), main="background plot")
orig_par <- par()

fig.new.1 <- c(grconvertX(c(0.25, 0.75), from="npc", to="ndc"),
               grconvertY(c(0.15, 0.40), from="npc", to="ndc"))
par(fig=fig.new.1, cex=.5, new=TRUE, mar=rep(0, 4))
plot(1:10, col="red", main="inset plot 1")
par(orig_par)

fig.new.2 <- c(grconvertX(c(0.25, 0.75), from="npc", to="ndc"),
               grconvertY(c(0.65, 0.90), from="npc", to="ndc"))
par(fig=fig.new.2, cex=.5, new=TRUE, mar=rep(0, 4))
plot(1:10, col="blue", main="inset plot 2")
par(orig_par)

points(seq(0,1,by=0.5), seq(0,1,by=0.5), pch=16, cex=2)
```
![image](https://user-images.githubusercontent.com/13220910/42704870-1b0d5714-8697-11e8-97f8-9092d1d410b4.png)

But you have to reset within each loop, or the inset will start to do this:
```
plot(NA, NA, xlim=c(0,1), ylim=c(0,1), main="background plot")
orig_par <- par()

fig.new.1 <- c(grconvertX(c(0.25, 0.75), from="npc", to="ndc"),
               grconvertY(c(0.15, 0.40), from="npc", to="ndc"))
par(fig=fig.new.1, cex=.5, new=TRUE, mar=rep(0, 4))
plot(1:10, col="red", main="inset plot 1")
# WE DON'T RESET HERE (SO NO RESET IN LOOP)

fig.new.2 <- c(grconvertX(c(0.25, 0.75), from="npc", to="ndc"),
               grconvertY(c(0.65, 0.90), from="npc", to="ndc"))
par(fig=fig.new.2, cex=.5, new=TRUE, mar=rep(0, 4))
plot(1:10, col="blue", main="inset plot 2")
par(orig_par)

points(seq(0,1,by=0.5), seq(0,1,by=0.5), pch=16, cex=2)
```
![image](https://user-images.githubusercontent.com/13220910/42704933-4eb17cf8-8697-11e8-883b-0b1823afa482.png)